File: htmlParser.hs
main :: IO ()
main = do
    contents <- readFile "file.html"
    scan contents 0 [] ([],[0,0,0])

scan :: (Num t, Num a1, Num a2, Eq t, Eq a1, Eq a2) => [Char] -> t -> [Char] -> ([a2], [a1]) -> IO ()
scan [] open string (stack, [a,b,c])
    | open /= 0 = errorWithoutStackTrace "Missing '>' symbol in file."
    | not (null stack) = errorWithoutStackTrace "Detected unclosed tag(s) in HTML file."
    | a == 0 = errorWithoutStackTrace "HTML file does not contain any valid html tags."
    | c /= 2 = errorWithoutStackTrace "The file must contain a single body section and be surrounded by <html>...</html>."
    | otherwise = putStrLn "HTML file successfully parsed."

scan (x:xs) open string (stack, [a,b,c])
    | x == '<' = if open /= 0 
        then errorWithoutStackTrace "Consecutive '<' symbols seen."
        else scan xs 1 [] (stack, [a,b,c])
    
    | x == '>' = if open /= 1 
        then errorWithoutStackTrace "No matching '<' symbol seen for '>'."
        else scan xs 0 string (parse string stack [a,b,c])

    | x /= '>' && open == 1 = scan xs 1 (x:string) (stack, [a,b,c])
    | otherwise = scan xs open string (stack, [a,b,c])


-- Since cons is used to build the tag string, the string is in reverse order
-- Although the function could be called with the other parameter also manually written in reverse order, it would reduce readability.
-- Due to this, I have decided to use the reverse function despite a slight performance impact.
compare' :: String -> String -> Bool
compare' tag string = x == string && (length tag == num || y == ' ')
    where
    num = length string
    a = reverse tag      
    x = take num a
    y = a !! num


-- a = number of open tags detected
-- b = number of head tags detected
-- c = number of body tags detected
parse :: (Num a1, Num a2, Eq a1, Eq a2) => String -> [a2] -> [a1] -> ([a2], [a1])
parse tag [] [a,b,c]
    | compare' tag "html" = if a /= 0
        then errorWithoutStackTrace "<html> tag must be the first tag in the document and surround all other tags."
        else ([1], [1, b, c])
    
    | otherwise = errorWithoutStackTrace ("Invalid tag name detected: " ++ reverse tag)

parse tag (y:ys) [a,b,c]
    | compare' tag "/html" = if y /= 1
        then errorWithoutStackTrace "</html> in wrong position"
        else (ys, [a,b,c])

    | compare' tag "head" = if y /= 1 
        then errorWithoutStackTrace "<html> tag must directly precede <head> tag."
        else if b /= 0 
            then errorWithoutStackTrace "Multiple <head> tags detected."
            else if c /= 0 
                then errorWithoutStackTrace "<head> tag should precede <body> tag"
                else (2:(y:ys), [a+1,b+1,c])

    | compare' tag "/head" = if y /= 2 
        then errorWithoutStackTrace "</head> tag placed incorrectly."
        else (ys, [a,b,c])

    | compare' tag "body" = if y /= 1 
        then errorWithoutStackTrace "<html> tag must surround and precede <body> tag."
        else if c /= 0 
            then errorWithoutStackTrace "Multiple <body> tags detected."
            else (3:(y:ys), [a+1,b,c+1])

    | compare' tag "/body" = if y /= 3 
        then errorWithoutStackTrace "</body> tag in wrong position."
        else (ys, [a,b,c+1])

    | compare' tag "title" = if y /= 2 
        then errorWithoutStackTrace "<head> tag must surround and directly precede the <title> tag."
        else (4:(y:ys), [a+1,b,c])

    | compare' tag "/title" = if y /= 4 
        then errorWithoutStackTrace "</title> tag in wrong position."
        else (ys, [a,b,c])

    | c /= 1 = errorWithoutStackTrace "<body> tag must surround all relevant tags."

    | compare' tag "h1" = (5:(y:ys), [a+1,b,c])
    | compare' tag "/h1" = if y /= 5 
        then errorWithoutStackTrace "</h1> tag in wrong position."
        else (ys, [a,b,c])

    | compare' tag "h2" = (6:(y:ys), [a+1,b,c])
    | compare' tag "/h2" = if y /= 6 
        then errorWithoutStackTrace "</h1> tag in wrong position."
        else (ys, [a,b,c])

    | compare' tag "h3" = (7:(y:ys), [a+1,b,c])
    | compare' tag "/h3" = if y /= 7 
        then errorWithoutStackTrace "</h3> tag in wrong position."
        else (ys, [a,b,c])

    | compare' tag "p" = if y == 8 
        then errorWithoutStackTrace "<p> tag cannot be nested inside a <p> tag."
        else (8:(y:ys), [a+1,b,c])

    | compare' tag "/p" =  if y /= 8 
        then errorWithoutStackTrace "</p> tag in wrong position."
        else (ys, [a,b,c])

    | compare' tag "ul" = (9:(y:ys), [a+1,b,c])
    | compare' tag "/ul" = if y /= 9 
        then errorWithoutStackTrace "</ul> tag in wrong position."
        else (ys, [a,b,c])

    | compare' tag "li" = (10:(y:ys), [a+1,b,c])
    | compare' tag "/li" = if y /= 10 
        then errorWithoutStackTrace "</li> tag in wrong position."
        else (ys, [a,b,c])

    | compare' tag "a" = (11:(y:ys), [a+1,b,c])
    | compare' tag "/a" = if y /= 11 
        then errorWithoutStackTrace "</a> tag in wrong position."
        else (ys, [a,b,c])

    | compare' tag "div" = if y == 8 
        then errorWithoutStackTrace "<div> tag cannot be nested inside a <p> tag."
        else (12:(y:ys), [a+1,b,c])

    | compare' tag "/div" = if y /= 12 
        then errorWithoutStackTrace "</div> tag in wrong position."
        else (ys, [a,b,c])

    --ignore br and hr tags and make no change
    | compare' tag "br" || compare' tag "hr" = (y:ys, [a,b,c])
    | otherwise = errorWithoutStackTrace ("Invalid tag name detected: " ++ reverse tag)

File: htmlParser.c
#include <stdio.h>

#define MAXLINELENGTH 400


int stack[100];
int pointer = 0;
int tagCounter = 0;
int body = 0;
int head = 0;


// compare function to compare strings up to a certain limit
// the limit has been included as a parameter instead of being automatically calculated to provide a small performance
// boost.
int compare(const char *s1, const char *s2, int lim)
{
    for (int i=0; i<lim; i++)
    {
        if (s1[i] != s2[i])
        {
            return 0;
        }
    }
    if (s1[lim] != '\0' && s1[lim] != ' ')  // allow an optional space for tag attributes
    {
        return 0;
    }
    return 1;
}

int parse(char *tag)
{
    if (compare(tag, "html", 4))
    {
        if (tagCounter != 0 && pointer != 0)
        {
            printf("<html> tag must be the first tag in the document and surround all other tags.\n");
            return 0;
        }
        stack[pointer] = 1;
        pointer++;
        tagCounter++;
        return 1;
    }
    else if (compare(tag, "/html", 5))
    {
        if (stack[pointer-1] == 1)
        {
            stack[pointer-1] = 0;
            return 1;
        }
        printf("</html> tag in wrong position\n");
        return 0;
    }
    else if (pointer <= 0)
    {
        printf("<html> tag must be the first tag in the document and surround all other tags.\n");
        return 0;
    }
    else if (compare(tag, "head", 4))
    {
        if (stack[pointer-1] != 1)
        {
            printf("<head> tag placed outside of <html> tag\n");
            return 0;
        }
        if (head != 0)
        {
            printf("Multiple <head> tags detected.\n");
            return 0;
        }
        if (body > 0)
        {
            printf("<head> tag should precede <body> tag.\n");
            return 0;
        }
        stack[pointer] = 2;
        pointer++;
        tagCounter++;
        head++;
        return 1;
    }
    else if (compare(tag, "/head", 5))
    {
        if (stack[pointer-1] == 2)
        {
            pointer--;
            return 1;
        }
        printf("</head> tag placed incorrectly.\n");
        return 0;
    }
    else if (compare(tag, "body", 4))
    {
        if (stack[pointer-1] != 1)
        {
            printf("<html> tag must surround and precede <body> tag.\n");
            return 0;
        }
        if (body != 0)
        {
            printf("Multiple <head> tags detected.\n");
        }
        stack[pointer] = 3;
        pointer++;
        tagCounter++;
        body++;
        return 1;
    }
    else if (compare(tag, "/body", 5))
    {
        if (stack[pointer-1] == 3)
        {
            pointer--;
            body++;
            return 1;
        }
        printf("</body> tag in wrong position.\n");
        return 0;
    }
    else if (compare(tag, "title", 5) == 1)
    {
        if (stack[pointer-1] != 2)
        {
            printf("<head> tag must surround and directly precede the <title> tag.\n");
            return 0;
        }
        stack[pointer] = 4;
        pointer++;
        tagCounter++;
        return 1;
    }
    else if (compare(tag, "/title", 6))
    {
        if (stack[pointer-1] == 4)
        {
            pointer--;
            return 1;
        }
        printf("</title> tag in wrong position.\n");
        return 0;
    }
    else if (body != 1)
    {
        printf("<body> tag must surround all relevant tags.\n");
        return 0;
    }
    else if (compare(tag, "h1", 2))
    {
        stack[pointer] = 5;
        pointer++;
        tagCounter++;
        return 1;
    }
    else if (compare(tag, "/h1", 3))
    {
        if (stack[pointer-1] == 5)
        {
            pointer--;
            return 1;
        }
        printf("</h1> tag in wrong position.\n");
        return 0;
    }
    else if (compare(tag, "h2", 2))
    {
        stack[pointer] = 6;
        pointer++;
        tagCounter++;
        return 1;
    }
    else if (compare(tag, "/h2",3))
    {
        if (stack[pointer-1] == 6)
        {
            pointer--;
            return 1;
        }
        printf("</h2> tag in wrong position.\n");
        return 0;
    }
    else if (compare(tag, "h3", 2))
    {
        stack[pointer] = 7;
        pointer++;
        tagCounter++;
        return 1;
    }
    else if (compare(tag, "/h3", 3))
    {
        if (stack[pointer-1] == 7)
        {
            pointer--;
            return 1;
        }
        printf("</h3> tag in wrong position.\n");
        return 0;
    }
    else if (compare(tag, "p", 1))
    {
        if (stack[pointer-1] == 8)
        {
            printf("<p> tag cannot be nested inside a <p> tag.\n");
            return 0;
        }
        stack[pointer] = 8;
        pointer++;
        tagCounter++;
        return 1;
    }
    else if (compare(tag, "/p", 2))
    {
        if (stack[pointer-1] == 8)
        {
            pointer--;
            return 1;
        }
        printf("</p> tag in wrong position.\n");
        return 0;
    }
    else if (compare(tag, "ul", 2))
    {
        stack[pointer] = 9;
        pointer++;
        tagCounter++;
        return 1;
    }
    else if (compare(tag, "/ul", 3))
    {
        if (stack[pointer-1] == 9)
        {
            pointer--;
            return 1;
        }
        printf("</ul> tag in wrong position.\n");
        return 0;
    }
    else if (compare(tag, "li", 2))
    {
        stack[pointer] = 10;
        pointer++;
        tagCounter++;
        return 1;
    }
    else if (compare(tag, "/li", 3))
    {
        if (stack[pointer-1] == 10)
        {
            pointer--;
            return 1;
        }
        printf("</li> tag in wrong position.\n");
        return 0;
    }
    else if (compare(tag, "a", 1))
    {
        stack[pointer] = 11;
        pointer++;
        tagCounter++;
        return 1;
    }
    else if (compare(tag, "/a", 2))
    {
        if (stack[pointer-1] == 11)
        {
            pointer--;
            return 1;
        }
        printf("</a> tag in wrong position.\n");
        return 0;
    }
    else if (compare(tag, "div", 3))
    {
        if (stack[pointer-1] == 8)
        {
            printf("<div> tag cannot be nested inside a <p> tag.\n");
            return 0;
        }
        stack[pointer] = 12;
        pointer++;
        tagCounter++;
        return 1;
    }
    else if (compare(tag, "/div", 4))
    {
        if (stack[pointer-1] == 12)
        {
            pointer--;
            return 1;
        }
        printf("</div> tag in wrong position.\n");
        return 0;
    }
    else if (compare(tag, "br", 2) || compare(tag, "hr", 2))
    {
        return 1;
    }
    else
    {
        printf("Invalid tag name detected: %s\n", tag);
        return 0;
    }

}

int checkValidity(int opened)
{
    if (opened)
    {
        printf("Missing '>' symbol in file.\n");
        return 0;
    }
    if (pointer != 1)
    {
        printf("Detected unclosed tag(s) in HTML file.\n");
        return 0;
    }
    if (tagCounter == 0)
    {
        printf("HTML file does not contain any valid html tags.\n");
        return 0;
    }
    if (body != 2)
    {
        printf("The file must contain a single body section and be surrounded by <html>...</html>.\n");
        return 0;
    }
    return 1;
}

int getInputText()
{
    char tag[MAXLINELENGTH];
    int i = 0;
    FILE *inputFile = fopen("file.html", "r");
    int opened = 0;
    while (!feof(inputFile))
    {
        char c = (char) fgetc(inputFile);
        if (c == '<')
        {
            if (!opened)
            {
                opened = 1;
            }
            else
            {
                printf("Consecutive '<' symbols seen.\n");
                return 0;
            }
        }
        else if (c == '>')
        {
            if (opened)
            {
                opened = 0;
                tag[i] = '\0'; // add null byte char to denote the end of the string
                int valid = parse(tag);
                if (!valid) { return 0;}
                i = 0;
            }
            else
            {
                printf("No matching '<' symbol seen for '>'.\n");
                return 0;
            }
        }
        else if (opened)
        {
            tag[i] = c;
            i++;
        }
    }
    fclose(inputFile);
    return checkValidity(opened);
}

int main() {
    if (getInputText())
    {
        printf("HTML file successfully parsed.\n");
    }
}
